### PR TITLE
Add sue

### PR DESCRIPTION
**Pull request by theAkito**
## What `sue` does
Executes a program as a user different from the user running sue. The target program is exec'ed which means, that it replaces the sue process you are using to run the target program. This simulates native tools like su and sudo and uses the same low-level POSIX tools to achieve that, but eliminates common issues that usually arise, when using those native tools.

## Why `sue` exists
Maintainable alternative to ncopa/su-exec, which is the better tianon/gosu. This one is far better (higher performance, smaller size), than the original gosu, however it is far easier to maintain, than su-exec, which is written in plain C.

### CODE REVIEW

**Comment by mash1t** [comment]
Hey @theAkito,
thank you for your contribution.
Please correct me if i'm wrong but i can not see a connection between sue and Docker.
As this is a Awesome Docker list all projects do have to do with Docker in one way or another.

**Comment by theAkito** [comment]
@mash1t 

As described in several places, like [here for example](https://github.com/tianon/gosu#warning), this kind of tool is not very suited for interactive use. Its primary use-case is for inside-Docker environments. Using it outside of Docker most of the time means you are misusing the tool. Additionally, I do not know any real production use-case where such a tool is used outside of Docker. It is only used within Docker.

Therefore, it should qualify as a Docker tool.

**Comment by veggiemonk** [comment]
Hi @theAkito @mash1t 

The reason to have this list is to gather projects related to building and using docker. Of course, being "related" is loosely defined but a good judgment implies here.

Most people running containers will run into issues with access rights regarding users. The section clearly specifies that it lists: "Tools and applications that are either installed inside containers or designed to be run as a sidecar".
Therefore, init programs (or startup scripts) are included.

@theAkito I think @mash1t has a point though. I don't see anywhere in the README of the project a paragraph showing "how to use it with a container". An example Dockerfile would go a long way.

**Comment by theAkito** [comment]
> The section clearly specifies that it lists: "Tools and applications that are either installed inside containers or designed to be run as a sidecar".
> Therefore, init programs (or startup scripts) are included.

It only makes sense to run it in a Docker environment, as it is now. It is not suited to be run outside of Docker. Therefore, it is a de facto Docker utility, as explained above and as referenced in many other places.



> @theAkito I think @mash1t has a point though. I don't see anywhere in the README of the project a paragraph showing "how to use it with a container". An example `Dockerfile` would go a long way.

I already explained extremely detailed and explicitly, what the program is, why it exists and how to use it. You just have to scroll down. The only thing "missing" is that I didn't use the word "Docker" in the README.md, because it is implied anyway.
When you are talking about people arriving to rescue you from a fire, I think everybody understands that you are talking about firefighters.



> Also, you mention that sue is better but it lacks any kind of data or benchmarks.

The whole point of `su-exec` is that it is by far not as bloated as the original `gosu`. So the selling point of it is its size (about 10kb).
However, the cost of this optimizations is lower maintainability. So I created a middle ground between optimization and maintainability.
BUT, as already mentioned, all this is already explained in all the READMEs, etc. Everything I said here was clear from the start. I am just repeating myself and others.



> Finally, it says it easier to maintain than others written in C. Why is that easier to maintain?

Have you ever written something in C? Writing in C is like writing a novel like Lord of the Rings in an old runic and half unknown or still encrypted language, which you have to finish a whole major for, to even remotely grasp parts of it. So it is not suited for open source projects, as much as pretty much any other language in the Top 10 is, right now. Literally any other modern production ready language is better to maintain than C. I am starting to think this question was meant to be ironic and perhaps I fell for that.

> It is written in Nimble which I barely heard of. Maybe you can put that in the README :)

It is written in Nim, which is a language as simple to understand, read and use, as Python-like languages, yet it is almost as fast as C and you can use C directly in it, through its FFI, without any blocking or weird behaviour (I wouldn't even call C a "foreign" language in this situation).
You can see all that when you have a look at the source code. Importing a C procedure in Nim has more or less the same difficulty as importing the former in C++, which is extremely easy, therefore more maintainable already from this narrow aspect of the whole arsenal of reasons why a language like Nim is _just better_.

Basically, when using Nim, you can use plain C, if you are forced to, and reach performance stats similar to C's, but all the other parts of the source code will be as readable and maintainable as a slow and simple scripting language.

All that said, as long as `su-exec` and `gosu` is in the list, then `sue` has a place there, as well, indubitably.

**Comment by theAkito** [comment]
@mash1t 
@veggiemonk 

I have added an explicit and detailed explanation of how `sue` is supposed to be used with Docker:
https://github.com/theAkito/sue#where

I have no idea how to make it more explicit and straight to the face, without making that explanation a parody of itself.

If you use this program outside of Docker, you most likely are doing something wrong. That's why it belongs to the list of tools that are Docker related.
I use the word _related_ on purpose here, since **any** Docker related program, that only should be used with Docker, could theoretically be used in some weird way outside of Docker. So if you want to administrate every item on the list as strictly as possible by the rules, you shouldn't have any items in an "awesome-docker" list, as any program could theoretically be used in a non-Docker environment.

**Review by veggiemonk — APPROVED** [review]

**Comment by theAkito** [comment]
Thanks to everyone, I am happy I could get this resolved. 👍🏻
